api/src/main/java/com/netflix/iceberg/expressions/UnboundPredicate.java

```diff
@@ -132,4 +134,36 @@ public Expression bind(Types.StructType struct, boolean caseSensitive) {
     }
     return new BoundPredicate<>(op(), new BoundReference<>(struct, field.fieldId()), lit);
   }
+
+  private Types.NestedField findNestedField(Types.StructType struct, String expressionFieldPath,
+    boolean caseSensitive) {
+
+    String[] nestedFields = expressionFieldPath.split("\\.");
+    String lastFieldInPath = nestedFields[nestedFields.length - 1];
+
+    Types.StructType subField = struct;
+    int i=0;
+
+    Type nextFieldType = caseSensitive ? subField.field(nestedFields[i]).type() :
+      subField.caseInsensitiveField(nestedFields[i]).type();
+
+    if (!(nextFieldType instanceof Types.StructType)) {
+      // we don't handle non-struct nested fields yet
+      return null;
+    }
+
+    while(nextFieldType instanceof Types.StructType && i < nestedFields.length) {
+
+      subField = caseSensitive ? subField.field(nestedFields[i]).type().asStructType() :
+        subField.caseInsensitiveField(nestedFields[i]).type().asStructType();
+
+      i++;
+
+      nextFieldType = caseSensitive ? subField.field(nestedFields[i]).type() :
+        subField.caseInsensitiveField(nestedFields[i]).type();
+    }
+
+    return subField.field(lastFieldInPath);
+
+  }
```

Comment:
Now that we need to recurse, I wonder if it makes sense to reuse the indexes available in `TypeUtils`: https://github.com/apache/incubator-iceberg/blob/0c9c63140e838875dc8cc52a57be2c8f24ad9975/api/src/main/java/com/netflix/iceberg/types/TypeUtil.java#L78-L84 so that this code simplifies to:

```java
Integer idx = indexByName.get(expressionFieldPath);
Types.NestedField field = indexById.get(idx);
```
That code doesn't check for non-struct parents; we'd have to perhaps create a custom Indexer. Also we would need to calculate the index lazily and perhaps cache it ( As in the `Schema`, see https://github.com/apache/incubator-iceberg/blob/master/api/src/main/java/com/netflix/iceberg/Schema.java#L59-L71 )

Reply:
Doesn't look like TypeUtils.indexByName & TypeUtils.indexById are used when reading data. It's used by the Manifest reader to index on manifest file statistics data.

Reply:
I guess you are suggesting we now start using this. i dunno what impact that would have though. do you see a major benefit to it? I can look into it.

Reply:
> I guess you are suggesting we now start using this.

Right.

> do you see a major benefit to it?

Code reuse. But it would only make sense if `findNestedField()` is called many times. The first call to `findNestedField()` would build the indexes and consult them. Then any subsequent call to `findNestedField()` would just consult the cached Maps.

Reply:
Right. seems like it's being called many times but by different expression evaluators. A cache to lookup field name to Types.NestedField would help. Good call.  will add one in.

Reply:
@xabriel I'm now creating an index that keeps ids to accessors and then use accessors to reach type and fields. access to this index is lazy as well.